internal/apidiff/apidiff.go

```diff
@@ -0,0 +1,284 @@
+	ClientLibraryType string `json:"client_library_type"`
+	DocsURL           string `json:"docs_url"`
+	ReleaseLevel      string `json:"release_level"`
+}
+
+type differ struct {
+	root     string
+	manifest map[string]manifestEntry
+	results  map[string]string
+	errs     map[string]error
+}
+
+func newDiffer() (*differ, error) {
+	root, err := os.Getwd()
+	if err != nil {
+		return nil, err
+	}
+	return &differ{
+		root:    root,
+		results: map[string]string{},
+		errs:    map[string]error{},
+	}, nil
+}
+
+func (d *differ) diffModules() error {
+	goCloud, err := cloneGoCloud()
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(goCloud)
+
+	for imp, entry := range d.manifest {
+		// Only diff stable clients.
+		if entry.ReleaseLevel != "ga" {
+			continue
+		}
+
+		// Prepare module directory paths relative to the repo root.
+		pkg := strings.TrimPrefix(imp, rootMod+"/")
+		baseModDir := goCloud
+		modDir := d.root
+
+		// Manual clients are also submodules, so we need to run apidiff in the
+		// submodule.
+		if entry.ClientLibraryType == "manual" {
+			baseModDir = path.Join(baseModDir, pkg)
+			modDir = path.Join(modDir, pkg)
+		}
+
+		// Create apidiff base from repo remote HEAD.
+		base, err := d.writeBase(baseModDir, imp, pkg)
+		if err != nil {
+			d.errs[imp] = err
+			continue
+		}
+
+		// Diff the current checked out change against remote HEAD base.
+		out, err := d.diff(modDir, imp, pkg, base)
+		if err != nil {
+			d.errs[imp] = err
+			continue
+		}
+
+		if out != "" {
+			d.results[imp] = out
+		}
+	}
+
+	return nil
+}
+
+func (d *differ) writeBase(baseModDir, imp, pkg string) (string, error) {
+	if err := cd(baseModDir); err != nil {
+		return "", err
+	}
+
+	base := path.Join(baseModDir, "pkg.master")
+	out, err := execE("apidiff", "-w", base, imp)
+	if err != nil && !isSubModErr(out) {
+		return "", err
+	}
+
+	// If there was an issue with loading a submodule, change into that
+	// submodule directory and try again.
+	if isSubModErr(out) {
+		parent := d.manualParent(imp)
+		if parent == pkg {
+			return "", fmt.Errorf("unable to find parent module for %q", imp)
+		}
+		if err := cd(parent); err != nil {
+			return "", err
+		}
+		out, err := execE("apidiff", "-w", base, imp)
+		if err != nil {
+			return "", fmt.Errorf("%s: %s", err, out)
+		}
+	}
+	return base, nil
+}
+
+func (d *differ) diff(modDir, imp, pkg, base string) (string, error) {
+	if err := cd(modDir); err != nil {
+		return "", err
+	}
+	out, err := execE("apidiff", "-incompatible", base, imp)
+	if err != nil && !isSubModErr(out) {
+		return "", err
+	}
+	if isSubModErr(out) {
+		parent := d.manualParent(imp)
+		if parent == pkg {
+			return "", fmt.Errorf("unable to find parent module for %q", imp)
+		}
+		if err := cd(parent); err != nil {
+			return "", err
+		}
+		out, err = execE("apidiff", "-w", base, imp)
+		if err != nil {
+			return "", fmt.Errorf("%s: %s", err, out)
+		}
+	}
+
+	return out, err
+}
+
+func (d *differ) loadManifest() error {
```

Comment:
@codyoss just realizing that we cannot use this script on go-genproto because there is no repo-metdata file there... :/